Correct highlighting for asymmetric matchers

grosto opened the pull request:
<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. The two fields below are mandatory. -->

<!-- Please remember to update CHANGELOG.md in the root of the project if you have not done so. -->

## Summary

I've opened this PR for https://github.com/facebook/jest/issues/6184 and https://github.com/facebook/jest/issues/6170.

From what I could find, there are two cases where Jest shows a confusing diff:
1. Object properties, which pass asymmetric matchers, are represented as failed fields.
2. Objects, which are checked against `objectContaining` or `arrayContaining` ,  are fully shown in diff instead of only showing the subset of an object.

@pedrottimark I will be waiting for your input regarding what and how :)
I have already read your comment in https://github.com/facebook/jest/issues/7027#issuecomment-423837270. 

## Test plan

I have added several snapshot tests. I am not sure if this is the best way to illustrate the issue with tests, so the suggestions are highly welcome :)

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI. -->

pedrottimark commented:
Yes, thank you for these snapshot tests as baseline so we can discuss desired report.

What do you think about moving `expect.objectContaining` down one level like hypothetical Flux standard action:

```js
jestExpect({type: 'whatever', payload: {a: 'a', b: 'b', c: 'c', d: 'd'}}).toEqual(
  {
    type: 'whatever',
    payload: expect.objectContaining({
      a: 'x',
      b: expect.any(String),
    }),
  },
)
```

For your info, a solution to this problem will probably have to be powerful enough to solve another problem of **deleted** or especially **inserted** properties:
* during first phase of TDD
* incorrect assumptions about received value

Even though we mitigated the problem with a more efficient diff algorithm, I think Jest can do even better by understanding when properties are “equal” versus deleted or inserted.

Here are some important special cases when changes include both deleted and inserted props:
* key remained the same but value changed
* key changed but value remained the same

Although object properties are the first priority, here is an open-ended thought question for more meaningful diff: realistic scenarios when **array items** don’t match up. For example, React has `key` prop to decide when a collection changes, is each child the same versus deleted or inserted versus changed.

grosto commented:
Here is a small update for now.

To get some inspiration and make sure that I am not gonna try to reinvent the wheel. I have been reading several codebases for the past week(`deep-diff`, `concordance`, `jest-diff` and some others related to diffing two objects). 

Since yesterday I started designing a solution. I mostly use objects to reason about the problem, but I want to make sure that in the future the solution will be able to handle other data structures.

So far, I think I am on track, but I might bother you for some advice later this week.

SimenB commented:
Fantastic @grosto! Looking forward to it 🙂

thymikee commented:
Hey @grosto, anything we can help with to get this over the line? :)

grosto commented:
Hey @thymikee,
I have not had any time to dedicate to this for the past two weeks. Hopefully, I will continue working on it later this week. If this is a high priority, maybe it's better if someone else will take over :)

pedrottimark commented:
@grosto no worries, same for me.

WeiAnAn commented:
Hey, I have some idea to solve this problem.

When `jest-diff` compare object, we iterate the object first.
If there is asymmetric matcher, check it is match or not.
If match, copy the asymmetric matcher to corresponding attribute.

Example: 
```js
const copyB = {...b};
Object.keys(a).forEach(key => {
  if (a[key].$$typeof === Symbol.for('jest.asymmetricMatcher')) {
    if (a[key].asymmetricMatch(copyB[key])) {
      copyB[key] = Object.assign(a[key], {});
    }
  }
});
```
And using this new object to diff.
Result: 
![result](https://i.imgur.com/6WxUTf9.png)

grosto commented:
I am back on this!

Sorry, I started a new job and really could not find a spare second. :)

I will need a day or two to remember the codebases and gather my thoughts again. I will get back with some ideas after that.

grosto commented:
After having a fresh look at the problem, I was not really fond of my ideas from a month ago 😅. So I decided to think a bit more about it.

In order to provide a better visual representation of data differences,  we have to change our approach from serialize-then-diff to diff-then-serialize. As already mentioned by @pedrottimark.

Below is some idea(In a very very naive Typescript):
```ts
const INSERTED = 'INSERTED';
const DELETED = 'DELETED';
const EQUAL = 'EQUAL';
const UPDATED = 'UPDATED';
const TYPE_EQUAL = 'TYPE_EQUAL';

type DiffResult =
  | typeof INSERTED
  | typeof DELETED
  | typeof EQUAL
  | typeof UPDATED
  | typeof TYPE_EQUAL;

interface PrimitiveValuesDiff {
  diffResult: DiffResult;
  path?: string;
  rhsValue?: PrimitiveValue;
  lhsValue?: PrimitiveValue;
  // there will be more fields needed for sure
}

interface ComplexValuesDiff {
  lhsConstructorName?: string;
  rhsConstructorName?: string;
  path?: string;
  diffResult: DiffResult;
  propertyDiffs: (ComplexValuesDiff | PrimitiveValuesDiff)[];
  // there will be more fields needed for sure
}

// example of two object diff
const expected = {
  a: 1,
  b: {
    c: expect.any(Number),
    d: 2,
    e: 3,
  },
};

const received = {
  a: 1,
  b: {
    c: 1,
    e: 4,
  },
};

diffComplexValues(expected, received, options);

{
  lhsConstructorName: 'Object',
  rhsConstructorName: 'Object',
  diffResult: TYPE_EQUAL,
  /* type equality is sufficient for complex values */
  propertyDiffs: [
    {
      path: 'a',
      rhsValue: 1,
      lhsValue: 1,
      diffResult: EQUAL,
    },
    {
      path: 'b',
      lhsConstructorName: 'Object',
      rhsConstructorName: 'Object',
      diffResult: UPDATED,
      propertyDiffs: [
        {
          path: 'c',
          diffResult: EQUAL,
          lhsValue: expect.any(Number),
          rhsValue: 1,
        },
        {
          path: 'd',
          diffResult: DELETED,
          lhsValue: 2,
        },
        {
          path: 'e',
          diffResult: UPDATED,
          lhsValue: 3,
          rhsValue: 4,
        },
      ],
    },
  ],
};


`
- Expected
+ Received

  Object {
    "a": 1,
    "b": {
      "c": 1,
  -   "d": 2,
  -   "e": 3,
  +   "e": 4,
    }
  }
`
```

The ComplexValuesDiff will be sufficient to produce a string representation of data differences/commonalities. With some extra fields, this data structure can be also used to describe diffs of Arrays, Maps, and Sets. 

Important Note: I am not suggesting to create a big diff object and then pass it to another function to generate a diff string. We are probably going to create a diff string as we are traversing an object. The only reason why I made these types is that it's easier for me to illustrate the idea.

It is crucial that diff function is consistent with the equality one. We can extract the smaller functions from current `equals`. Then use these small functions to create a consistent equality and diff functions.
The new diff function can be used instead of `compareObjects` function in `jest-diff`, or it could be part of `expect` package. 

I am still trying to solve `expect.objectContaining`. It only has `asymmetricMatch` method, which does the comparison and returns a boolean. The traversal of the object happens inside the asymmetric matcher, so there is no way to see which fields pass/failed from outside this class. I have some ideas on how to solve this, but I am not sure about them. Any suggestions are welcome :)

Glad to hear any feedback :)

jeysal commented:
@grosto Super excited that you're working on tackling this long-standing design limitation!

Quick question: What's the difference between `UPDATED` and `TYPE_EQUAL`? Paths `[]` and `['b']` look like they should be the same case?

`expect.*` will probably supply functions that generate a custom diff specific to the matcher, right? If so, `expect.objectContaining` could generate an `EQUAL` node with the `received` value for properties that match or were unspecified, and `UPDATED` for mismatches?

pedrottimark commented:
Thank you for sharing your thoughts! 

> I am not suggesting to create a big diff object and then pass it to another function

Great minds think alike: Scott Hovestadt is working on streaming interface for Jest reporters

> I am still trying to solve `expect.objectContaining`

Yes, asymmetric matchers were a puzzle to me. Lack of API came up recently https://github.com/facebook/jest/issues/8295#issuecomment-483157466

> It is crucial that diff function is consistent with the equality one …

Yes indeed, this paragraph is brilliant

Yes, let’s start with descriptive data structure to prototype enough parts to illustrate a few of the most desired improvements. I experimented with tuples, but hard to debug when I messed up.

> Then use these small functions to create a consistent equality and diff functions

A question in my mind is how closely to couple the small functions to compare and format:

* loosest coupling is format functions match according to `lhsConstructor` and `rhsConstructor` properties, similar to serializer plugins which have a `test` method
* tightest coupling is compare function assigns format function as a property

I think the solution will be looser, but a vague idea for asymmetric matchers is to be able to provide formatter as optional property of the comparison data structure.

Which reminds me to throw out for discussion possible end goal for packaging:

* move equality (after deep cleaning of edge cases and breaking changes) into separate package like `jest-equal`
* create corresponding comparison into new package in Jest monorepo
* create corresponding formatting into new package in Jest monorepo

pedrottimark commented:
Forgot to say about consistent functions, as comparison corresponds to equality, format functions borrow the best from `pretty-format` and `jest-diff` and overcome their limitations.

In your example, be able to display the change lines for updated primitive value:

* property key in dim color
* property values in green versus red

```diff
-   "e": 3,
+   "e": 4,
```

Also I think default `--no-expand` option will mean something other than number of adjacent lines.

pedrottimark commented:
While thinking about formatting that is possible given comparison, here is a use case to consider

Highlight which properties in received object match expected object for negative `toMatchObject`

Example picture baseline at left and **improved at right**:

<img width="1000" alt="toMatchObject true" src="https://user-images.githubusercontent.com/11862657/56929309-428b0300-6aa7-11e9-83e6-a8549b0ecc34.png">

grosto commented:
@jeysal Thank you for your input!

> Quick question: What's the difference between UPDATED and TYPE_EQUAL?

`UPDATED` means that primitive values are not equal. When it comes to complex values, we do not need to know if they are deep equal or not. All we care about is if their type is equal. if it is, we continue checking their keys, otherwise we can mark them as `UNEQUAL` and move on. I see how this is confusing with current types. 

@pedrottimark Thank you for your feedback. I am glad that we have the same vision and thank you for providing more problems to keep in mind while solving this problem.

> Highlight which properties in received object match expected object for negative toMatchObject

I have been thinking about the interface to create a subset diff for `toMatchObject` /`objectContaining` and in general, make diff more customizable.

Currently, I am working on extracting smaller functions from `equals`  and how to compose new diff and equality with them.

pedrottimark commented:
While trying to write a minimal formatter, here are thoughts about the descriptive data structure:

* distinguish “complex” from “basic” comparison objects by presence or absence of `propertyDiffs`
* for `lhsConstructorName` and `rhsConstructorName` replace string with actual object as value
* and then have consistent keys in “complex” and “basic” comparison object as `lhs` and `rhs` or even consider as terse as `a` and `b` following convention in `jasmineUtils.ts` and `utils.js`

jeysal commented:
> if it is, we continue checking their keys, otherwise we can mark them as `UNEQUAL` and move on.

But do you not want to mark them properly once we've finished recursing into them? If `TYPE_EQUAL` could mean `EQUAL` and `UPDATED` it e.g. won't allow you to collapse a part of the diff where you know everything is equal.

pedrottimark commented:
@grosto @jeysal

* Maybe reverse the reasoning so `TYPE_UNEQUAL` is value for **non** comparable complex values, like Object and typed array, for which you might short-circuit the comparison?
* An interesting case is comparison of basic versus complex value, when it seems like `UPDATED` and no need for `propertyDiffs` but the formatter might display properties of the complex value (this is evidence why the object instance is needed as a value in comparison object)
* In my rough prototyping, a complex comparison function kept local `propertyDiffs` array, but returned it as property value only if `diffResult` was `UPDATED` not for `EQUAL`

These cases are evidence against what I suggested about absence of `propertyDiffs` as a hint ;)

When no relevant prop diffs for complex value, possible formatting convention could include ellipsis:
* `{…}` for plain object or `[…]` for plain array
* `MyClass {…}` for derived class or `Int32Array […]` typed array

SimenB commented:
Is this fixed now that we've landed #9257?

grosto commented:
Hi, 

The scope of this PR was intended to be larger than just asymmetric matchers, it's more about changing the jest's diff solution. Is this still something that jest's team considers doing? I have not had free time to contribute to open source for the past months, but I would love to come back to this problem. Also, I think this PR contains valuable context if somebody else would pick it up.

SimenB commented:
>it's more about changing the jest's diff solution. Is this still something that jest's team considers doing?

If it's improved/more usable, we're always open to it 🙂 It's at least been documented now: https://github.com/facebook/jest/blob/master/packages/jest-diff/README.md

---

Happy to keep this open, I just wondered if it might have been solved 👍

grosto commented:
Hello,

I have finally found some time for this. Sorry for keeping it for so long. I have pushed partial implementation with tests and README.md. Will be waiting for the feedback.

grosto commented:
To clarify I am waiting for a feedback on this one to continue.

jeysal commented:
I'm so sorry you haven't received feedback on this yet :(
I'm *very* interested in this topic (in fact I considered exploring an impl of this myself) but atm I'm barely getting my work on other things (the `babel-plugin-jest-hoist` rewrite, the prettier-less snapshots) done :cry:
Please don't be discouraged by this, I will get to it sooner or later :upside_down_face: and it's totally fine if you want to wait for feedback before proceeding of course.

grosto commented:
No worries. Thanks for replying.

The rest of the work will be based on this foundation and I want to make sure it's a good one, before I continue. So I would rather wait for feedback for now.

mattvb91 commented:
I think this would currently help me figuring out a diff I cant make heads or tails off due to it currently outputting the whole objects instead of the actually difference between them.

jeysal commented:
> I will get to it sooner or later

Yeah this was later rather than sooner.
Fortunately, as this is a new package intended to completely replace existing ones, it does not get out of date :see_no_evil: 
So, @pedrottimark has not been seen on Jest for a while, but maybe we can get this over the line still?
The code is obviously a lot and some pasted so it's hard to review without anyone working full-time on Jest other than scanning for obvious open points, but I suppose if we can just do a run of the `expect toStrictEqual` test values on this and make sure every non-equal value pair generates a diff on this and every equal pair generates no diff, then that's probably a good enough indicator that it's solid.
Plus maybe some of us can link it into some Jest work projects, use it for a while.
This is just way too much work and way too important to leave hanging like this, and the `DiffObject` structure looks like exactly what we need :see_no_evil: wdyt @SimenB

grosto commented:
> Yeah this was later rather than sooner.

No worries. I am here still to get this to the finish line, I was just not sure if there was still interest. 
`diffComp.test.ts` runs some tests to make sure that the behavior of this diff and old diff match in most cases. 
I understand that there is a lot of code did you manage to read README file? it describes what's missing and how everything is currently implemented. Let me know if you have any immediate feedback from reading it.

jeysal commented on packages/jest-deep-diff/README.md:
````diff
@@ -0,0 +1,122 @@
+Example:
+
+```ts
+const line: Line = {type: Inserted, prefix: '"a": ', sufix: ',', val: {b: 1}};
````
Yeah, when reading this example (in particular the `{b: 1}` part at the end) I already kind of saw the question down there coming :sweat_smile:

To express what I understand in my words and make sure we're on the same page:
The original `DiffObject` is a tree structure, while this `Line` does not have further `Line`s in its `val` - it is flat. So these `Line`s can only represent the leaves of the `DiffObject` tree, meaning in your example the outer `{a: ...}` (which is UPDATED, not a leaf) is handled differently from the inner `{a: ...}` (which is UNEQUAL_TYPE, a leaf).

At the risk of suggesting something you had already thought about and know is problematic:
Flattening the tree seems to be the problem - I think we want to traverse a full `DiffObject` tree that goes all the way down into all nested structures.
Does diffing `{}` vs `{a: {b: 1}}` result in tree 1 (from what I understand it currently does):
```
UPDATED Object
  INSERTED a: Object...
```
or tree 2:
```
UPDATED Object
  INSERTED a: Object...
    INSERTED b: 1
```
If we have tree 2 (we still go all the way into an `INSERTED` property, it's just that every child will necessarily also be `INSERTED`), then the printer can do a recursive traversal pass through the tree, increasing indentation whenever it descends into child diffs, and really all that the `INSERTED`/`DELETED` kinds do is modify the sign that the printer prints at the start of the line.
`EQUAL` vs `UPDATED` would be almost the same for the printer really, only difference I can imagine is that the printer must never abbreviate an `UPDATED` node, it always has to print it out to get to the actual differences inside.
`UNEQUAL_TYPE` is a bit awkward, I think it would have to go and instead be a pair of `DELETED` and `INSERTED`, because what kind would the child diffs of an `UNEQUAL_TYPE` be?
Let me know what you think of this approach. I think so far the assumption of "when a new property appears (or a property has a different type), mark the whole thing different in some form and move on" has not really been challenged, but it seems implementing a printer has shown that it would be beneficial to have the WHOLE structure already analyzed so that the printer can just traverse through it - otherwise this problem occurs where `UPDATED` objects are treated as transparent by the printer and recursed into, while `INSERTED` objects are treated as opaque and just printed out as one big bag of characters, necessarily using different code for the two.

grosto replied:
It's an interesting approach. I am investigating these days whenever I have time.

From a high-level view, this approach would change the responsibilities of `DiffObject`, but probably it will help out formatter. I liked that `DiffObject` had no idea about the formatting part. It's a bit unclear to me what's the responsibility of `DiffObject` is now. 

From the implementation view, all nested structure diff functions require two values. `diffObjects(a, b, opts, diff)` and the same in formatter. I am working to changing things up. It's not that simple. A lot of things and types need to be changed around. I will update here if I have new insights. Currently, I am thinking that the plugin system could be a bit harder to implement.

jeysal replied:
I think the responsibility is still fairly clear, it marks the parts that were different with `kind`s. It's true that there would now be lots of subtrees that are entirely equal. I view it as:
Before, the maximum tree depth we would store for equal parts of the objects is 1.
Now, the maximum tree depth we would store for equal parts of the objects is `POSITIVE_INFINITY`.
The equal parts were always there, it's just we didn't bother to store deeper levels of them. We were going through them anyway when figuring out that they were actually equal at the start. But now we store them entirely, because it helps consumers of the diff object, like the printer.

grosto replied:
Some updates:
1. Flattening Inserted/Delete the values(I will refer to it as flatten) is a completely different thing than diffing two objects. Some code can be reused between them but it's not a lot. For example, in the case of objects, only `getObjectKeys` gets reused. I was thinking to use the idea of empty structures(`{}` for objects and `[]` for arrays) and diff against them in a or b position to get flattened inserted/deleted diffs. But not sure if there can be an empty structure for everything, mostly asymmetric matchers worried me. Might get back to that idea at some point.
2. Plugins will need to provide flatten functions just like diff function. For example, ``diffWithCustomDiffs`` is created in the main diff function with plugins and then passed down. The same would need to happen with flatten function. Those two points combined, I think that maybe it's best to add flatten as a separate step? It's an extra traversal but it's a bit awkward to carry both flatten and diff functions around.  
```ts

export function diffObjects(
  a: Record<Key, unknown>,
  b: Record<Key, unknown>,
  path: Path,
  diff: Function, 
  flatten: Function,
): DiffObject<Record<Key, unknown>> {
    // omited
    if (otherIndex >= 0) {
      childDiffs.push(diffFunc(a[key], b[key], key));
      bKeys[otherIndex] = null;
    } else {
      childDiffs.push(
        flatten({val: a[key], path: key, kind: Kind.DELETED, flatten}),
      );
    }
    // omited
}
```
So it with extra step it would be diff -> flatten -> format -> print. Not sure if it's worth it, but something I would be willing to try.

3. Switching UNEQUAL_TYPE with INSERTED and DELETED messes up models a bit. Instead of one DiffObject, the diff function now will have to return Array of DiffObjects, because inserted and deleted cannot be represented as one DiffObject. So far I am keeping it. When the formatter gets it, it just breaks it up into InsertedDiff and DeletedDiff and passes them back to itself. Since formatter returns an array of lines it composes well. Something like this.
```ts
const insertedDiff: DiffObject = {
  b: diffObj.b,
  childDiffs: diffObj.bChildDiffs,
  kind: Kind.INSERTED,
  path: diffObj.path,
};
const deletedDiff: DiffObject = {
  a: diffObj.a,
  childDiffs: diffObj.aChildDiffs,
  kind: Kind.DELETED,
  path: diffObj.path,
};
return [
  ...originalFormat(deletedDiff, context, options),
  ...originalFormat(insertedDiff, context, options),
];
```
In the future, formatter might just show the name of the constructor only when types are unequal or be smart about it in other ways.
```
-  "a": 3
+  "a": Object { ... }
```

4. Now asymmetricMatchers: what does the diff object of `expect.any(Object)` and `{ a: 1 }` looks like? Because they are equal, but we won't have flattened object for it. So far I am doing this, but the assumption that EQUAL means two objects have the same structure and will get flattened during the diffing process does not always stand.
``` js
 if (test(a)) {
    return a.asymmetricMatch(b)
      ? diff(b, b, path, memos) // Hacky way to get flattened equal object
      : flatten({ a, b, path, kind: TYPE_UNEQUAL }); 
 } 
```
5. Plugins might still have to provide serializer function, but not sure yet. Have to play around with react element plugin more.

Meanwhile, I will work to clean up the code(mostly types) and push it. But hopefully, this is enough to get the idea of some problems.

jeysal replied:
Hmmm, I'm not sure I follow. You are raising problems that occur when trying to flatten a diff object tree, but what I was trying to say is I would want to avoid even doing a flatten at all costs.
In my suggestion, the diff object tree would always have as much depth as the values we're comparing; it would include all parts of a and b, where a and b are equal, where a has a property that b doesn't have at all, and in any other case.
That tree would never be flattened, instead the printer would operate by traversing into the full-depth structure.

grosto replied:
Oops. Had a brain fart and went with it fully. `flatten` is not the right name but it does exactly what you described. It should be named `markRecursively` or `expand`. Just to make sure we are on the same page, this is the current behaviour.
```js
let a = {
  b: { c: 1 }
};
let b = {};


createDiffObject(a, b)
/*
DiffObject {
  "a": {
    "b": {
      "c": 1
    }
  },
  "b": {},
  "childDiffs": [
    {
      "a": {
        "c": 1
      },
      "childDiffs": [
      {
        "a": 1,
        "kind": "DELETED",
        "path": "c"
      }
      ],
      "kind": "DELETED",
      "path": "b"
    }
  ],
  "kind": "UPDATED"
}
*/
```

jeysal replied:
Okay, that clarifies a lot.

1)
Would it be possible to use `undefined` as that empty value for any type?
As in, `diff(null, {p:1})` is an `UNEQUAL_TYPE`, but `diff(undefined, {p:1})` is an `INSERTED`.
That way, I could imagine the step recursing into the property to be a call to `diff(undefined, 1)`, also generating an `INSERTED` using the same code, hopefully?
The reason to use `undefined` and not any other value of course being that undefined is already what represents a missing property in `a`, which also generates an `INSERTED` for that property in `b`.

3)
Okay, if we are keeping it, and if my 1) could work, recursing into an `UNEQUAL_TYPE` could also mean diffing against `undefined`, also generating `INSERTED` child diff objects, which sounds fine

4)
I too thought it would mean inserting the concrete successfully matched object as both a and b, like in your code example. Are you anticipating any problems, any requirements we cannot fulfill with this approach?

grosto replied:
1. There are couple of problems with that:
First, We cannot use `undefined` because it will treat missing key and key which has value of `undefined` the same way. 
Second, I tried the same approach but with different special type, but it just passed down the problem(and functions) to the next guy and makes code very hard to follow. In general diffing algorithms and traversing algorithms are quite different in our case and there is very little elegant reuse that can be achieved. I feel like I tried several approaches already and always hit the dead end. 
4. For now it's not a problem, it might be but we can worry about it later. None of this traversal changes are catastrophic and can always be reverted.

I will try to push code ASAP now so you also can get clearer idea

jeysal replied:
Hmm, couldn't it still be represented differently to make the strict equality `undefined` vs missing property that you mention work?
Consider the following pseudocode (of course simplified without things like `path`, also ignoring the other way around with `keysOnlyInA` and `DELETED`, ...) - what am I missing?
```js
const childDiffs = [];
let commonKeys /* = ... */;
let keysOnlyInB /* = ... */;
for (commonKey of commonKeys) {
  childDiffs.push(diff(a[commonKey], b[commonKey]))
}
for (keyOnlyInB of keysOnlyInB) {
  childDiffs.push({
    ...diff(undefined, b[keyOnlyInB]),
    kind: 'INSERTED',
  })
}
```

This is a really interesting conversation by the way :ok_hand: in general assertions and diffing are such an interesting topic :sweat_smile:

grosto replied:
I agree it's very interesting and it's great to hear your thought process about it too. 

I have been working on this whenever I had time. Pushing the newest changes so you can see them, but it's not final work.
This version is recursively traversing all complex objects and marks children as inserted/deleted.
Instead of `undefined` I went with Symbol which marks an empty value.  You can see the code in `complex/object.ts` in `diff` function. 

Additionally improved typing and rebased, which was some work considering how much changed.

I am planning to work on it more actively this week.
First I will try to make CI happy and then go through skipped tests(which are mostly for missing features) and try to make them pass.

Do you have any remarks/questions regarding overall architecture?

jeysal replied:
The union type does look cool indeed :+1: 
I would say overall architecture is mostly comprised of how `diff` traverses through data to build the diff object tree, and how `format` traverses through the diff object tree to print it, so I think we're aligned at this point :sweat_smile: 
Did you, while writing the code, feel like the approach I pushed for (that relies more heavily on tree-traversal) was helping against code duplication and feeling more natural? Or were there still awkward bits?
That would be interesting to me because of course my suggestions were all coming from quite a high level and I can't know for sure if they still work well when getting into the details implementing it

grosto replied:
I have tried and it has some quirks. Not saying no to it, but I wanted to implement some other types and see some diffs first. Since diff function API is mostly agreed on, how we implement it can be improved down the line.

jeysal reviewed (commented):
First of all let me say again this appears to be very well thought through and tested, again and again I am surprised how much you already have in place like the `diffComp.test` that I didn't even see at first because it was collapsed :see_no_evil: 
The main open point (things like extending support for more `Map`s etc. take effort of course, but probably aren't architecturally complex and have few unknowns left) seems to be the problem of printing `UPDATED` objects differently from `INSERTED` objects, so I've left a comment there to get the discussion going on how that duplication could be eliminated :upside_down_face: welcoming anyone, not just previous contributors, to join in on this!

grosto commented:
Now that architecture is somewhat stable(at least API of diff object), I wanted to implement some new thing and also fix some skipped tests. I also observed some interesting properties/tradeoffs of diff and then serialize approach.

Let's start with the pros:
I implemented expect.objectContaining and you can see the differences(the top one is old diff):
```TS
const a = {
      payload: {a: 'a', b: 'b', c: 'c', d: 'd'}, 
      type: 'whatever'
};
const b = {
    payload: expect.objectContaining({
        a: 'x',
        b: expect.any(String),
    }),
    type: 'whatever',
 };
```
This is field updated example. Inserted, Deleted and Unequal Type fields all have big improvements too.

![image](https://user-images.githubusercontent.com/19875911/116819567-ddc08280-ab70-11eb-94e6-666557049cfb.png)

We can also correctly understand inverse matches
```TS
const a = {
  payload: {a: 'a', b: 'b', c: 'c', d: 'd'}, 
  type: 'whatever'
};

const b = {
  payload: expect.not.objectContaining({
    a: 'x',
    b: expect.any(String),
  }),
  type: 'whatever',
};
```

![image](https://user-images.githubusercontent.com/19875911/116819627-25470e80-ab71-11eb-9fb0-c527644d468b.png)

```TS
const a = {
  payload: expect.not.objectContaining({
    a: 'a',
    b: expect.any(String),
  }),
  type: 'whatever',
};

const b = {
  payload: {a: 'a', b: 'b', c: 'c', d: 'd'},
  type: 'whatever',
};
```

![image](https://user-images.githubusercontent.com/19875911/116819667-49a2eb00-ab71-11eb-8f19-2c7a2b35a95b.png)

I also expect to see the big improvements on arrayContaining too. Have to implement it first.

Downsides:
Map diff is harder to implement because keys can have diffs too not just equality. So it means that we have to compare each key in the map to other keys until we find the closest match. For now, I am only comparing two keys with `Map.has()`.
It's easier for old diff to do this because just looks for differences in strings and it does not have any notion of data structures. To make it visual
```TS
const a = new Map([[{a: 1}, 2]]);
const b = new Map([[{a: 2}, 2]]);
```
![image](https://user-images.githubusercontent.com/19875911/116820034-3ee95580-ab73-11eb-9c1d-92e523d7ae56.png)

Currently, I am not traversing the complex keys of the map, so formatter cannot display them properly.  That's easily solvable. Bigger complexity is the diff algorithm.

The second downside is easier to see than to explain. I am not even sure if this is a downside, but there are some similarities that old diff picks up that we cannot pick up(easily at least)
```TS
const a = {
  searching: '',
  whatever: {
    a: 1,
  },
};

const b = {
  searching: '',
  whatever: [
    {
      a: 1,
    },
  ],
};
```

![image](https://user-images.githubusercontent.com/19875911/116820189-f0888680-ab73-11eb-86d4-efd95fb305e2.png)

Let me know what you think and I will continue working on this.
I still have to implement the remaining asymmetricMatchers, DOM and Set. Improve Map.  Add back React Element and finally add the same options as old diff had

jeysal commented:
I think the second 'downside' is totally okay, we don't need to show similarities that happen to be somewhere deep down in the structure even if what's wrapping them is totally different.
The map thing is annoying indeed, I totally didn't have it in my mind that you could have these kinds of keys to a map...

One thing struck me as odd:
![NotContaining example](https://user-images.githubusercontent.com/19875911/116819627-25470e80-ab71-11eb-9fb0-c527644d468b.png)
In this diff, the new impl generates a 'no visual difference'? Isn't that less than ideal?

grosto commented:
They are equal so there is no difference to show. Wording could be changed. I copied it the old diff i think

jeysal commented:
Are they? It's expecting an object not containing `b: any string`, and getting an object with `b: 'b'` isn't it?

grosto commented:
I even check in the test that they are equal.
https://github.com/facebook/jest/pull/7893/files#diff-93c714843afcedbd3f472f3dc8da523e81a2c23630a6e2e299fa80c4423aaac1R314

I get where you are coming from. I don't think that's how objectContaining behaves. It just checks if one object is subset of another.

jeysal commented:
Ah right yeah. Sorry. We even had a debate or some changes around how `.not.objectContaining` works, but apparently I had the wrong idea in my mind about how it works now after all :sweat_smile:

SimenB commented:
@grosto @jeysal hi! are you still interested in landing this PR? I've merged in `main`, but I don't know what else is missing

jeysal commented:
Funny, just yesterday I was thinking of this because I was myself experimenting with deep diffing algorithms.
I think this is still in the top 5 reworks that Jest needs, along with config, ESM, faster default transforms, etc.

I mean, LOL at this dumb diff that obviously knows nothing but string diffing:
![image](https://user-images.githubusercontent.com/16069751/189345789-5cf38ce9-0a92-4acd-8727-37f3b4f4127d.png)

Concordance (AVA) gets these cases better already for a long time.

---

Integrating it into `expect`, `pretty-format`, `jest-diff` as a behind the scenes implementation, that is I think where the majority of the remaining effort lies.
I would also want to hammer some `fast-check` tests at all the matchers using new and old impl to be really sure that passes and fails are unchanged (only formatting/diff should be rightfully changed).

In terms of implementation IIRC this was looking pretty good to me the last time I looked at it.
Since it is not yet actually used (it only implements the new package) there is little harm in merging it. Yes it will be dead code, sort of, for now. But if we ever continue this rework, this should be the code that we use.

jeysal commented:
As you know @SimenB I've been completely inactive on Jest this entire year, and aside from a blog post mostly inactive in 2021 too.

I'm currently taking a couple months break from work traveling etc. I've been thinking about what I want to do next. One of my thoughts was whether we could utilize some of that OpenCollective money to actually "employ" me, either just for a while until I start a more "classic" job again, or part-time next to some other job, so that we can get some of these big projects moving, get the issue tracker under control, etc.

Some food for thought...in about a month or so I will probably have to decide what to do next.

SimenB commented:
Oooh, I think it's a great idea if we could get somebody on top off the repo on a more consistent basis than my "I do what seems interesting, sometimes"! Let's discuss on discord
